src/Features/Core/Portable/Shared/Utilities/DeserializationConstructorCheck.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Shared.Utilities
+{
+    internal readonly struct DeserializationConstructorCheck
```

Comment:
is there a huge benefit to this type, versus just having an extension method on `Compilation` to do this?

Reply:
I hinted at this in the PR description:

> It cannot be an extension method as it caches symbols from the compilation for reuse across multiple calls.

"Cannot" isn't strictly true. Of course it can. But that would end up being three lookups per symbol, rather than three per compilation. Do you prefer an extension method here all the same?

Reply:
> "Cannot" isn't strictly true. Of course it can. But that would end up being three lookups per symbol, rather than three per compilation. Do you prefer an extension method here all the same?

I would.  Three lookups seems like nothing... especially since Compilations cache these calls to GetTypeByMetadataName.  Now, if we had perf data indicating a cost here, i would feel differently.

*That said*: i'm not opposed to this PR in this form.  So don't go changing it for me :)   I was just pointing out that a bunch of this felt like premature-optimization territory, and i'd prefer to avoid that as a general principle.  But, there's nothing wrong with this work at all, and you've gotten it this far, so i'm totally ok taking as-is :)

Reply:
I admit I hadn't looked at how the symbols were resolved. That does appear to be pretty inexpensive.

Would you define such an extension method on `ISymbolExtensions`?

@mavasani you mentioned porting this code in https://github.com/dotnet/roslyn/issues/32133#issuecomment-451255084. How do you see this?

Reply:
> Would you define such an extension method on ISymbolExtensions?

That would be reasonable :)

Reply:
Normally, we recommend caching well known types per-compilation in the analyzers, especially if only the analyzer needs a well known type. I am personally fine with all the three approaches here (your original approach, Sam's recommended approach and Cyrus' recommended approach) - we can always tune it if we see performance issues, which seem unlikely given GetTypeByMetadataName does cache.

Reply:
GetTypeByMetadataName only caches the most recently used 50 items. The current approach doesn't seem to have a real downside and saves those 50 slots for extension analyzers.